File: src/main/java/io/github/Earth1283/aptMc/commands/sub/ExportCommand.java
package io.github.Earth1283.aptMc.commands.sub;

import io.github.Earth1283.aptMc.AptMc;
import io.github.Earth1283.aptMc.commands.SubCommand;
import io.github.Earth1283.aptMc.managers.PackageManager;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import org.bukkit.command.CommandSender;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import com.google.gson.JsonObject;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.*;

public class ExportCommand extends SubCommand {

    public ExportCommand(AptMc plugin, PackageManager packageManager) {
        super(plugin, packageManager);
    }

    @Override
    public void execute(CommandSender sender, List<String> args, boolean dryRun) {
        String filename = args.isEmpty() ? "apt-manifest.yml" : args.get(0);
        if (!filename.endsWith(".yml")) filename += ".yml";
        File file = new File(plugin.getDataFolder(), filename);

        if (dryRun) {
            sender.sendMessage(plugin.getMessage("status.dry-run", Placeholder.unparsed("arg", "Export to " + filename)));
            return;
        }

        sendStatus(sender, plugin.getMessage("status.exporting", Placeholder.unparsed("arg", filename)));

        plugin.getServer().getScheduler().runTaskAsynchronously(plugin, () -> {
            Map<String, String> installed = packageManager.getInstalledPlugins();
            if (installed.isEmpty()) {
                sender.sendMessage(plugin.getMessage("status.export-empty"));
                return;
            }

            Map<String, JsonObject> versions;
            try {
                versions = resolveVersions(new ArrayList<>(installed.values()));
            } catch (Exception e) {
                sender.sendMessage(plugin.getMessage("status.export-failed", Placeholder.unparsed("error", "Failed to resolve versions: " + e.getMessage())));
                return;
            }

            FileConfiguration yaml = new YamlConfiguration();

            // Add Header
            yaml.set("project-details.title", "Server Export");
            yaml.set("project-details.author", sender.getName());
            yaml.set("project-details.created", System.currentTimeMillis());

            List<Map<String, String>> pluginList = new ArrayList<>();
            List<String> unrecognisedList = new ArrayList<>();
            int count = 0;

            for (Map.Entry<String, String> entry : installed.entrySet()) {
                String pluginFilename = entry.getKey();
                String sha1 = entry.getValue();

                if (versions.containsKey(sha1)) {
                    JsonObject v = versions.get(sha1);
                    String projectId = v.get("project_id").getAsString();
                    String verNum = v.get("version_number").getAsString();

                    Map<String, String> entryMap = new LinkedHashMap<>();
                    entryMap.put("file", pluginFilename);
                    entryMap.put("source", "modrinth:" + projectId + "/" + verNum);
                    pluginList.add(entryMap);
                    count++;
                } else {
                    unrecognisedList.add(pluginFilename);
                }
            }
            yaml.set("plugins", pluginList);
            if (!unrecognisedList.isEmpty()) {
                yaml.set("unrecognised", unrecognisedList);
            }

            // Config bundling
            sendStatus(sender, plugin.getMessage("exporting-configs")); // Fixed key name
            Map<String, Map<String, String>> configsMap = new HashMap<>();
            File pluginsDir = plugin.getDataFolder().getParentFile();
            
            // Count total files first for progress
            int totalFiles = countFilesRecursive(pluginsDir, pluginsDir);
            java.util.concurrent.atomic.AtomicInteger processedFiles = new java.util.concurrent.atomic.AtomicInteger(0);
            java.util.function.Consumer<Double> progressCallback = createProgressCallback(sender, "Exporting Configs");
            
            exportConfigsRecursive(pluginsDir, pluginsDir, configsMap, processedFiles, totalFiles, progressCallback);
            progressCallback.accept(1.0); // Ensure 100% at end
            
            List<Map<String, String>> configExportList = new ArrayList<>();
            for (Map.Entry<String, Map<String, String>> entry : configsMap.entrySet()) {
                String pluginName = entry.getKey();
                for (Map.Entry<String, String> fileEntry : entry.getValue().entrySet()) {
                   Map<String, String> cMap = new LinkedHashMap<>();
                   cMap.put("plugin", pluginName);
                   cMap.put("path", fileEntry.getKey());
                   cMap.put("data", fileEntry.getValue());
                   configExportList.add(cMap);
                }
            }
            yaml.set("configs", configExportList);


            try {
                yaml.save(file);
                sender.sendMessage(plugin.getMessage("status.export-success", Placeholder.unparsed("arg1", String.valueOf(count)), Placeholder.unparsed("arg2", file.getPath())));
                if (!unrecognisedList.isEmpty()) {
                    sender.sendMessage(plugin.getMessage("status.export-unrecognised-warning", Placeholder.unparsed("arg", String.valueOf(unrecognisedList.size()))));
                }
            } catch (IOException e) {
                sender.sendMessage(plugin.getMessage("status.export-failed", Placeholder.unparsed("error", e.getMessage())));
            }
        });
    }

    private int countFilesRecursive(File dir, File root) {
        int count = 0;
        File[] files = dir.listFiles();
        if (files == null) return 0;

        for (File f : files) {
            if (f.isDirectory()) {
                // Skip apt-mc data folder
                if (f.equals(plugin.getDataFolder())) continue;
                count += countFilesRecursive(f, root);
            } else {
                if (shouldExport(f)) {
                    count++;
                }
            }
        }
        return count;
    }

    private void exportConfigsRecursive(File dir, File root, Map<String, Map<String, String>> configs, java.util.concurrent.atomic.AtomicInteger processed, int total, java.util.function.Consumer<Double> callback) {
        File[] files = dir.listFiles();
        if (files == null) return;

        for (File f : files) {
           if (f.isDirectory()) {
                if (f.equals(plugin.getDataFolder())) continue;
                exportConfigsRecursive(f, root, configs, processed, total, callback);
           } else {
               if (shouldExport(f)) {
                   try {
                       byte[] content = new byte[(int) f.length()];
                       try (FileInputStream fis = new FileInputStream(f)) {
                           fis.read(content);
                       }
                       String b64 = Base64.getEncoder().encodeToString(content);
                       
                       // Determine plugin name from path relative to plugins dir
                       // Structure: plugins/PluginName/config.yml
                       String relPath = f.getPath().substring(root.getPath().length() + 1);
                       int firstSlash = relPath.indexOf(File.separator);
                       if (firstSlash != -1) {
                           String pluginName = relPath.substring(0, firstSlash);
                           String subPath = relPath.substring(firstSlash + 1);
                           
                           configs.computeIfAbsent(pluginName, k -> new HashMap<>()).put(subPath, b64);
                       }
                   } catch (Exception e) {
                       plugin.getLogger().warning("Failed to export config: " + f.getName());
                   }
                   
                   int p = processed.incrementAndGet();
                   if (total > 0 && p % 5 == 0) { // Update every 5 files to reduce spam
                        callback.accept((double) p / total);
                   }
               }
           }
        }
    }

    private boolean shouldExport(File f) {
        FileConfiguration config = plugin.getConfig();
        String mode = config.getString("export.filter-mode", "blacklist");
        List<String> extensions = config.getStringList("export.extensions");
        String name = f.getName();
        String ext = "";
        int i = name.lastIndexOf('.');
        if (i > 0) ext = name.substring(i + 1);

        if (mode.equalsIgnoreCase("whitelist")) {
             return extensions.contains(ext);
        } else {
             return !extensions.contains(ext);
        }
    }

}

File: src/main/kotlin/io/github/Earth1283/aptMc/commands/sub/ExportCommand.kt
package io.github.Earth1283.aptMc.commands.sub

import com.google.gson.JsonObject
import io.github.Earth1283.aptMc.AptMc
import io.github.Earth1283.aptMc.commands.SubCommand
import io.github.Earth1283.aptMc.managers.PackageManager
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import org.bukkit.command.CommandSender
import org.bukkit.configuration.file.YamlConfiguration
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.Base64
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

class ExportCommand(plugin: AptMc, packageManager: PackageManager) : SubCommand(plugin, packageManager) {

    private val flags = listOf("--include-all", "--compress", "--no-configs")

    companion object {
        private const val LARGE_FILE_THRESHOLD = 25L * 1024 * 1024
    }

    override fun onTabComplete(sender: CommandSender, args: List<String>): List<String> {
        val current = args.lastOrNull() ?: ""
        return flags.filter { it.startsWith(current, ignoreCase = true) && !args.dropLast(1).contains(it) }
    }

    override fun execute(sender: CommandSender, args: List<String>, dryRun: Boolean) {
        val remaining = ArrayList(args)
        val includeAll = remaining.removeIf { it.equals("--include-all", ignoreCase = true) }
        val compress = remaining.removeIf { it.equals("--compress", ignoreCase = true) }
        val noConfigs = remaining.removeIf { it.equals("--no-configs", ignoreCase = true) }

        val rawName = if (remaining.isEmpty()) "apt-manifest" else remaining[0]
        val baseName = rawName.removeSuffix(".yml").removeSuffix(".zip")
        val outputFile = File(plugin.dataFolder, baseName + if (compress) ".zip" else ".yml")

        if (dryRun) {
            sender.sendMessage(plugin.getMessage("status.dry-run", Placeholder.unparsed("arg", "Export to " + outputFile.name)))
            return
        }

        sendStatus(sender, plugin.getMessage("status.exporting", Placeholder.unparsed("arg", outputFile.name)))

        plugin.server.scheduler.runTaskAsynchronously(plugin, Runnable {
            val installed = packageManager.getInstalledPlugins()
            if (installed.isEmpty()) {
                sender.sendMessage(plugin.getMessage("status.export-empty"))
                return@Runnable
            }

            val versions: Map<String, JsonObject>
            try {
                versions = resolveVersions(ArrayList(installed.values))
            } catch (e: Exception) {
                sender.sendMessage(plugin.getMessage("status.export-failed", Placeholder.unparsed("error", "Failed to resolve versions: " + (e.message ?: ""))))
                return@Runnable
            }

            val yaml = YamlConfiguration()

            yaml.set("project-details.title", "Server Export")
            yaml.set("project-details.author", sender.name)
            yaml.set("project-details.created", System.currentTimeMillis())

            val pluginList = ArrayList<Map<String, String>>()
            val unrecognisedList = ArrayList<String>()
            var count = 0

            for ((pluginFilename, sha1) in installed) {
                val v = versions[sha1]
                if (v != null) {
                    val projectId = v.get("project_id").asString
                    val verNum = v.get("version_number").asString

                    val entryMap = LinkedHashMap<String, String>()
                    entryMap["file"] = pluginFilename
                    entryMap["source"] = "modrinth:$projectId/$verNum"
                    pluginList.add(entryMap)
                    count++
                } else if (includeAll) {
                    try {
                        val jarFile = File(packageManager.getPluginsDir(), pluginFilename)
                        val b64 = Base64.getEncoder().encodeToString(jarFile.readBytes())

                        val entryMap = LinkedHashMap<String, String>()
                        entryMap["file"] = pluginFilename
                        entryMap["source"] = "embedded"
                        entryMap["data"] = b64
                        pluginList.add(entryMap)
                        count++
                    } catch (e: Exception) {
                        plugin.logger.warning("Failed to embed unrecognised plugin: $pluginFilename")
                        unrecognisedList.add(pluginFilename)
                    }
                } else {
                    unrecognisedList.add(pluginFilename)
                }
            }
            yaml.set("plugins", pluginList)
            if (unrecognisedList.isNotEmpty()) {
                yaml.set("unrecognised", unrecognisedList)
            }

            if (!noConfigs) {
                sendStatus(sender, plugin.getMessage("status.exporting-configs"))
                val configsMap = HashMap<String, MutableMap<String, String>>()
                val pluginsDir = plugin.dataFolder.parentFile

                val totalFiles = countFilesRecursive(pluginsDir)
                val processedFiles = AtomicInteger(0)
                val progressCallback = createProgressCallback(sender, "Bundling configurations", "status.config-progress")

                exportConfigsRecursive(pluginsDir, pluginsDir, configsMap, processedFiles, totalFiles, progressCallback)
                progressCallback.accept(1.0)

                val configExportList = ArrayList<Map<String, String>>()
                for ((pluginName, fileMap) in configsMap) {
                    for ((path, data) in fileMap) {
                        val cMap = LinkedHashMap<String, String>()
                        cMap["plugin"] = pluginName
                        cMap["path"] = path
                        cMap["data"] = data
                        configExportList.add(cMap)
                    }
                }
                yaml.set("configs", configExportList)
            }

            try {
                if (compress) {
                    val yamlName = baseName + ".yml"
                    ZipOutputStream(FileOutputStream(outputFile)).use { zos ->
                        zos.putNextEntry(ZipEntry(yamlName))
                        zos.write(yaml.saveToString().toByteArray(Charsets.UTF_8))
                        zos.closeEntry()
                    }
                } else {
                    yaml.save(outputFile)
                }
                sender.sendMessage(plugin.getMessage("status.export-success", Placeholder.unparsed("arg1", count.toString()), Placeholder.unparsed("arg2", outputFile.path)))
                if (unrecognisedList.isNotEmpty()) {
                    sender.sendMessage(plugin.getMessage("status.export-unrecognised-warning", Placeholder.unparsed("arg", unrecognisedList.size.toString())))
                }
                if (!compress && outputFile.length() > LARGE_FILE_THRESHOLD) {
                    sender.sendMessage(plugin.getMessage("status.export-large-file-warning"))
                }
            } catch (e: IOException) {
                sender.sendMessage(plugin.getMessage("status.export-failed", Placeholder.unparsed("error", e.message ?: "")))
            }
        })
    }

    private fun countFilesRecursive(dir: File): Int {
        var count = 0
        val files = dir.listFiles() ?: return 0

        for (f in files) {
            if (f.isDirectory) {
                if (f == plugin.dataFolder) continue
                count += countFilesRecursive(f)
            } else {
                if (shouldExport(f)) {
                    count++
                }
            }
        }
        return count
    }

    private fun exportConfigsRecursive(
        dir: File,
        root: File,
        configs: MutableMap<String, MutableMap<String, String>>,
        processed: AtomicInteger,
        total: Int,
        callback: Consumer<Double>
    ) {
        val files = dir.listFiles() ?: return

        for (f in files) {
            if (f.isDirectory) {
                if (f == plugin.dataFolder) continue
                exportConfigsRecursive(f, root, configs, processed, total, callback)
            } else {
                if (shouldExport(f)) {
                    try {
                        val content = ByteArray(f.length().toInt())
                        FileInputStream(f).use { fis ->
                            fis.read(content)
                        }
                        val b64 = Base64.getEncoder().encodeToString(content)

                        val relPath = f.path.substring(root.path.length + 1)
                        val firstSlash = relPath.indexOf(File.separatorChar)
                        if (firstSlash != -1) {
                            val pluginName = relPath.substring(0, firstSlash)
                            val subPath = relPath.substring(firstSlash + 1)

                            configs.computeIfAbsent(pluginName) { HashMap() }[subPath] = b64
                        }
                    } catch (e: Exception) {
                        plugin.logger.warning("Failed to export config: " + f.name)
                    }

                    val p = processed.incrementAndGet()
                    if (total > 0 && p % 5 == 0) {
                        callback.accept(p.toDouble() / total)
                    }
                }
            }
        }
    }

    private fun shouldExport(f: File): Boolean {
        val config = plugin.getConfig()
        val mode = config.getString("export.filter-mode", "blacklist")
        val extensions = config.getStringList("export.extensions")
        val name = f.name
        var ext = ""
        val i = name.lastIndexOf('.')
        if (i > 0) ext = name.substring(i + 1)

        return if (mode.equals("whitelist", ignoreCase = true)) {
            extensions.contains(ext)
        } else {
            !extensions.contains(ext)
        }
    }
}
